Source/Convai/Private/ConvaiUtils.cpp

```diff
@@ -404,8 +442,89 @@ namespace
 			return nullptr;
 		}
 	}
+
+	bool DecompressUSoundWave(USoundWave* soundWave, TSharedPtr<SongBufferData>& Out_SongBufferData)
+	{
+		FAudioDevice* audioDevice = GEngine ? GEngine->GetMainAudioDeviceRaw() : nullptr;
+
+		if (!audioDevice || !soundWave || soundWave->GetName() == TEXT("None"))
+			return false;
+
+		bool breturn = false;
+
+		// Ensure we have the sound data. Compressed format is fine.
+		soundWave->InitAudioResource(audioDevice->GetRuntimeFormat(soundWave));
+
+		// Create a decoder for this audio. We want the PCM data.
+		ICompressedAudioInfo* AudioInfo = audioDevice->CreateCompressedAudioInfo(soundWave);
+
+		// Decompress complete audio to this buffer
+		FSoundQualityInfo QualityInfo = { 0 };
+#if ENGINE_MAJOR_VERSION == 4
+		if (AudioInfo->ReadCompressedInfo(soundWave->ResourceData, soundWave->ResourceSize, &QualityInfo))
+		{
+			Out_SongBufferData = TSharedPtr<SongBufferData>(new SongBufferData(
+				QualityInfo.SampleDataSize, QualityInfo.NumChannels, QualityInfo.Duration, QualityInfo.SampleRate));
+
+			// Decompress all the sample data into preallocated memory now
+			AudioInfo->ExpandFile(Out_SongBufferData->RawPCMData.GetData(), &QualityInfo);
+
+			breturn = true;
+		}
+#else
+		FAudioDevice* AudioDevice = GEngine->GetMainAudioDeviceRaw();
+		if (AudioDevice)
+		{
+			FName format = AudioDevice->GetRuntimeFormat(soundWave);
+			soundWave->InitAudioResource(format);
+		}
+
+		const uint8* ResourceData = soundWave->GetResourceData();
+		uint32 ResourceSize = soundWave->GetResourceSize();
+
+		if (!ResourceData || ResourceSize <= 0)
+		{
+			return breturn;
+		}
+
+		if (AudioInfo->ReadCompressedInfo(ResourceData, ResourceSize, &QualityInfo))
+		{
+			Out_SongBufferData = TSharedPtr<SongBufferData>(new SongBufferData(
+				QualityInfo.SampleDataSize, QualityInfo.NumChannels, QualityInfo.Duration, QualityInfo.SampleRate));
+
+			// Decompress all the sample data into preallocated memory now
+			AudioInfo->ExpandFile(Out_SongBufferData->RawPCMData.GetData(), &QualityInfo);
+
+			breturn = true;
+		}
+#endif
+		// Clean up.
+		delete AudioInfo;
+
+		return breturn;
+	}
 };
 
+TArray<uint8> UConvaiUtils::ExtractPCMDataFromSoundWave(USoundWave* SoundWave, int32& OutSampleRate)
+{
+	TArray<uint8> PCMData;
+
+	if (!SoundWave)
+	{
+		UE_LOG(LogTemp, Warning, TEXT("SoundWave is null!"));
+		return PCMData;
+	}
+
```

Comment:
We need to check here at line:517 if the SoundWave is already not compressed

example:
	if (SoundWave->RawPCMDataSize > 0)
	{
		PCMData.Append(SoundWave->RawPCMData, SoundWave->RawPCMDataSize);
		OutSampleRate = SoundWave->GetSampleRateForCurrentPlatform();
	}
	else
	{
		TSharedPtr<SongBufferData> SongBuffer;
		if (DecompressUSoundWave(SoundWave, SongBuffer) && SongBuffer.IsValid())
		{
			PCMData = SongBuffer->RawPCMData;
			OutSampleRate = SongBuffer->BufferInfo.SampleRate;
		}
	}